pkg/agent/route/route_linux.go

```diff
@@ -1330,60 +1327,69 @@ func (c *Client) AddClusterIPRoute(svcIP net.IP) error {
 	}
 
-	// Generate a route with the new destination CIDR and install it.
+	// Generate a route with the new ClusterIP CIDR and install it.
 	newClusterIPCIDRMask, _ := newClusterIPCIDR.Mask.Size()
 	route := generateRoute(newClusterIPCIDR.IP, newClusterIPCIDRMask, gw, linkIndex, scope)
 	if err = netlink.RouteReplace(route); err != nil {
 		return fmt.Errorf("failed to install new ClusterIP route: %w", err)
 	}
-	// Store the new destination CIDR.
+
+	// Store the new ClusterIP CIDR and the new generated route to serviceRoutes. Then the calculated route can be restored
+	// when it was deleted since members of serviceRoutes are synchronized periodically.
 	if isIPv6 {
 		c.clusterIPv6CIDR = route.Dst
+		c.serviceRoutes.Store(clusterIPv6CIDRKey, route)
 	} else {
 		c.clusterIPv4CIDR = route.Dst
+		c.serviceRoutes.Store(clusterIPv4CIDRKey, route)
 	}
 	klog.V(4).InfoS("Created a route to route the ClusterIP to Antrea gateway", "route", route, "ClusterIP", svcIP)
 
 	// Collect stale routes.
 	var staleRoutes []*netlink.Route
 	if curClusterIPCIDR != nil {
-		// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled.
+		// If the current ClusterIP CIDR exists, the route with the current ClusterIP CIDR should be uninstalled since
+		// a new route with a newly calculated ClusterIP CIDR has been installed.
 		route.Dst = curClusterIPCIDR
 		staleRoutes = []*netlink.Route{route}
 	} else {
-		// If current destination CIDR is nil, which means that Antrea Agent has just started, then all existing routes
-		// whose destination CIDR contains the first ClusterIP should be uninstalled, except the newly installed route.
-		// Note that, there may be multiple stale routes prior to this commit. When upgrading, all stale routes will be
-		// collected. After this commit, there will be only one stale route after Antrea Agent started.
 		routes, err := c.listIPRoutesOnGW()
 		if err != nil {
 			return fmt.Errorf("error listing ip routes: %w", err)
 		}
+		// If current ClusterIP CIDR doesn't exist, which means that Antrea Agent has just started, and the route for
+		// ClusterIP generated in the last round should be deleted.
 		for i := 0; i < len(routes); i++ {
-			if routes[i].Gw.Equal(gw) && !routes[i].Dst.IP.Equal(svcIP) && routes[i].Dst.Contains(svcIP) {
+			// Skip the route whose destination CIDR only has one address.
+			ones, _ := routes[i].Dst.Mask.Size()
+			if ones == ipv4AddrLength || ones == ipv6AddrLength {
+				continue
+			}
+			// Find stale route.
+			if routes[i].Dst.Contains(svcIP) && routes[i].Gw.Equal(gw) {
 				staleRoutes = append(staleRoutes, &routes[i])
+				// TODO: add a break here. Some old versions of Antrea have issue https://github.com/antrea-io/antrea/issues/3131.
+				//  There might be multiple stale routes and they should be removed too. Without the issue, there should
+				//  be only one stale route, and a break should be added here to avoid unnecessary iteration.
```

Comment:
This is only executed once, I think no problem to iterate all routes to ensure thorough cleanup, regardless of how stale routes were produced and without the need to concern how many old versions it can be compatible with.

Reply:
OK, I'll rephrase the comment like the following:

```go
// Find stale routes. Note that, some old versions of Antrea have issue https://github.com/antrea-io/antrea/issues/3131.
// There might be multiple stale routes, and all of them should be removed. Without the issue, there should
// be only one stale route.
if routes[i].Dst.Contains(svcIP) && routes[i].Gw.Equal(gw) {
     staleRoutes = append(staleRoutes, &routes[i])
}
```